### ISSUE TITLE
[tracking] Experimental WebAssembly multithreading support

### ISSUE THREAD

**Issue by lambdageek** [dotnet/runtime#68162]
This is a tracking issue for adding experimental support for WebAssembly multi-threading to the browser-wasm runtime pack.

On the [branch](https://github.com/steveisok/runtime/commits/wasm-threading-eventpipe):
- [ ] put all modifications under conditionals
   - The runtime build should be under `/p:WasmEnableThreads=true`
   - ref assemblies should use `FeatureWasmThreads` in project files and `FEATURE_WASM_THREADS` in C# code to conditionally include `[UnsupportedOSPlatform("browser")]` guards around multi-threading APIs, and include the ref assembly in `Microsoft.NET.WebAssembly.Threading`
   - CoreLib uses `FeatureWasmThreads` and `FEATURE_WASM_THREADS` to define `System.Thread.IsThreadStartSupported`.  All other code should check `System.Thread.IsThreadStartSupported` to throw a PNSE.  There should not be `[UnsupportedOSPlatform("browser")]` guards in CoreLib
   - EventPipe and diagnostics should use `FeatureWasmDiagnostics` and `FEATURE_WASM_DIAGNOSTICS` (and a constant?).
 - [ ] Separate `browser-mt` and `browser-mt-eventpipe` samples.  
 - [ ] Land the branch changes in `main`
 
Experimental support in `main`:
- [ ] Build artifacts for the `/p:WasmEnableThreads=true` should go into a separate output folder. In other words it should be possible to build with `/p:WasmEnableThreads=false` in the same tree and without the build artifacts interfering with each other.
- [ ] The `browser-wasm` runtime pack should pack both single- and multi-threaded builds.  The multithreaded build artifacts should go into an `mt` subfolder.
- [ ] The `Microsoft.NET.WebAssembly.Threading.nupkg` ref assemblies package should be published to the dotnet7 feed
- [ ] The browser-wasm SDK should support a `WasmEnableThreads` property that uses the `mt` subfolder from the runtime pack for the runtime artifacts and that adds a `<PackageReference Include="Microsoft.NET.WebAssembly.Threading" IncludeAssets="compile" PrivateAssets="nonoe" />` package

EventPipe in `main`:
- [ ] Some kind of runtime build with multi-threading enabled internally, but not for the user APIs - EventPipe should be enabled.
- [ ] Save runtime artifacts from the EventPIpe build in an `ep` subfolder in the runtime pack
- [ ] browser SDK changes to use the `ep` artifacts if a property is set.
- [ ] JavaScript bindings to the EventPipe Session API - start stop sessions from JS, saving .nettrace to the VFS.  Some convenience API to extract and download the .nettrace file

Issues:
- [ ] Publish `dotnet.worker.js`
- [ ] Need a hook for frameworks to identify the location of `dotnet.js`.  Emscripten relies on `mainScriptUrlOrBlob` property on the module on the main threads to be set to the URL of `dotnet.js` so that it can locate `dotnet.worker.js`.  Frameworks like Blazor sometimes rename `dotnet.js` and/or move the wasm artifacts into subdirectories.  Provide a hook for them to tell the runtime where to find the files.
- [ ] protect workers from unhandled JS exceptions.  If user code uses `setTimeout` to queue work on the workers and that code throws unhandled exceptions, we should not bring down the whole runtime in a disorganized manner.
- [ ] single-threaded VFS.  Track Emscripten progress on the new VFS and upgrade when it's available.
- [ ] `Thread.Start().Wait` pattern  on the browser threads will deadlock. Emscripten relies on the browser thread's eventloop to start threads (unless we use a dedicated "main" thread when building).  After a thread start, the browser thread's eventloop must run in order for new threads to start.

Ecosystem:
- [ ] We need some way for package authors to provide single threaded and multi-threaded versions of native assets

**Comment by TheXenocide** [comment]
Ooh, just came to check in on how things were going on the mono wasm side of this and came across this tracking issue. Nice to see this being tracked on the dotnet/runtime side now. It's good to see that the browser-wasm RID was the final answer on that particular design challenge, but I remember how much discussion there was about trying to avoid adding additional runtimes unnecessarily. Trying to make sense of the current state of these tickets, are we expecting that the browser-wasm runtime will be multi-purpose and have some convention for choosing subtypes, rather than defining new RIDs for multi-threaded and perf-trace package variants? Part of me feels like the multi-thread variant would be better served by its own RID if there are plans to support both (fundamentally different) runtimes for any extended period (e.g. reference assemblies/contracts can expose different supported surface areas at compile time rather than throwing runtime exceptions).

Is there a sense of how this will work for packaging and compile MSBuild conditions? It seems like WasmEnableThreads is rather narrow and probably only impacts the application build, not so much for libraries, NuGet packaging, etc. unless the desired result is for the compiler to be able to generate all variants automatically, but it seems like there would be significant design implications there, where performance and a variety of other factors would want to be able to provide different implementations for the different environments (e.g. concurrency-safe collection types could ignore concurrency without MT but might need to leverage platform-specific no-lock behaviors, etc.) Understandable if these aren't answered yet, just trying to get a sense of how these things intend to be differentiated by component developers between the runtime and consuming developers.

I ask here mainly because it seems there are multiple possible threads that seem appropriate. Feel free to move this comment or direct me where best to move the discussion. Otherwise, glad to see there's already some experimental support running now.

**Comment by sajjadarashhh** [comment]
what happend if this feature not support with browser?

**Comment by knight1219** [comment]
How, if any, effect for this have on Task based threading? If one creates multiple HttpClient.GetAsync tasks and schedules them using Task.WhenAll, will all tasks run in parallel on their own threads? Does each task need the ConfigureAwait(false) to be set?

If that is the case and you are sharing this code between Blazor WASM and Blazor Hybrid, are there any repercussions in MAUI by doing that?

**Comment by kg** [comment]
Threading allows tasks to run on multiple threads, yes. Things that interact with browser APIs like HTTP are a more complex scenario, because browser APIs have thread affinity (think the nightmare of COM apartments, but worse)

**Comment by baywet** [comment]
I haven't, I was expecting the team here to revert with a workaround or a fix. I believe the key is hardcoded and I wasn't able to find extensibility points to impact the bundling process.

**Comment by czdietrich** [comment]
Any update on the `js-module-threads` issue?
It would be great to know how to work around this even if it currently needs manual steps.
So we could gather experience with the multithreading environment in Blazor in early development phases.

**Comment by curiousdannii** [comment]
This [ASP.net](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-7-rc-2/) blog has a very basic threading example, but I haven't been able to find one that includes `JSImport`s. I keep getting `ES6 module main.js was not imported yet, please call JSHost.Import() first.` errors.

Are threads and `JSImport`s meant to be working, or do I need to wait for another .NET release?

**Comment by lambdageek** [comment]
> Are threads and `JSImport`s meant to be working, or do I need to wait for another .NET release?
> 
> Edit: looks like I need to wait for #84489.

That's exactly right. It's not supported yet, but I'm working on it.

I'm curious to know more about your use case. 

Note that with #84489, only threads started with `Task.Run` will be able to JSImport async functions that return promises. 

Just starting a thread with `new Thread()` and calling some JSImported promise and then calling `.Wait()` or `.Result` won't be supported (those synchronous  blocking wait/result calls preventnus from returning to the JS event loop in the webworker, so JS promises never get a chance to resolve).

**Comment by curiousdannii** [comment]
> I'm curious to know more about your use case.

I have an app I'm trying to port, which calls a blocking function in a non-managed dll:

```
        [DllImport("Glk")]
        internal static extern void glk_select(ref Event ev);
```

To port it to WASM/JS it instead needs to call an async JS function (ignore that it no longer takes any arguments):

```
        [JSImport("glk_select", "main.js")]
        internal static partial Task glk_select();
```

That's pretty much it! I just need some way to block while I wait for the promise to resolve.

**Comment by ArcadeMode** [comment]
> Any update on the `js-module-threads` issue? It would be great to know how to work around this even if it currently needs manual steps. So we could gather experience with the multithreading environment in Blazor in early development phases.

Any news on this? I'm still seeing this today, not sure how to work around it.

**Comment by lambdageek** [comment]
> > Any update on the `js-module-threads` issue? It would be great to know how to work around this even if it currently needs manual steps. So we could gather experience with the multithreading environment in Blazor in early development phases.
> 
> Any news on this? I'm still seeing this today, not sure how to work around it.

Unfortunately Blazor support for threading is taking a bit longer than we would have liked. We're working on it, but it's not something that you can experiment with at this point. We do expect to have something in a later .NET 8 preview.

**Comment by ArcadeMode** [comment]
> Unfortunately Blazor support for threading is taking a bit longer than we would have liked. We're working on it, but it's not something that you can experiment with at this point. We do expect to have something in a later .NET 8 preview.

Thats a shame, thank you for your reply 😊

**Comment by pavelsavara** [comment]
I updated MT tracking issue https://github.com/dotnet/runtime/issues/85592

**Comment by tebeco** [comment]
hello,

Can we have this meta tracker updated somehow for net8 now that the release was done last november ?

![image](https://github.com/dotnet/runtime/assets/2266487/8838a611-f148-4965-bd56-eb9e25df18e7)

**Comment by pavelsavara** [comment]
> Anyone know if this meta tracker could updated for net8 ?

There were mostly low level steps toward multi-threading during Net8, but it's not supported in Net8.
For the current progress on Net9, please have look at https://github.com/dotnet/runtime/issues/85592

**Comment by tebeco** [comment]
> > Anyone know if this meta tracker could updated for net8 ?
> 
> There were mostly low level steps toward multi-threading during Net8, but it's not supported in Net8. For the current progress on Net9, please have look at #85592

<img width="1072" alt="image" src="https://github.com/dotnet/runtime/assets/2266487/a947741e-4bc5-44b8-8011-e2de4b6c2186">

This image is why I asked,
It was usefully showing net7.0 - RC2
and also showing net9.0 milestone, but is missing anything related net8.0 that being ultra small or bigger type of release

Now i understand that there's another one regarding Net9.0

The question is ... from an outsider point of you it would be nice to have this issue either
* renamed to clarify it's only net7 (probably not the case)
* closed if all done (probably not the case)
* updated every release (ie: net8.0: list of PR)
* list all possible tracker per release and each release issue summing it

I'm not trying to be picky, I'm just hardly able to try to follow up cross release "Epic size" feature when there's a lot of cross over that are not listed.
There's probably other ways i suppose but this ticket title seems like the perfect candidate to list the "per release"

**Comment by sbeji-fab** [comment]
Are there any updates regarding this feature? 
Its been silent for long time
